lib/AFP/Data/ByteString/Short/Word16.hs

```diff
@@ -0,0 +1,509 @@
+{-# LANGUAGE PackageImports #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE UnboxedTuples #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module AFP.Data.ByteString.Short.Word16 (
+  -- * Introducing and eliminating 'ShortByteString's
+  empty,
+  singleton,
+  pack,
+  unpack,
+
+  -- * Basic interface
+  snoc,
+  cons,
+  append,
+  last,
+  tail,
+  head,
+  init,
+  null,
+  length,
+
+  -- * Transforming ShortByteStrings
+  map,
+  intercalate,
+
+  -- * Reducing 'ShortByteString's (folds)
+  foldl',
+
+  -- ** Special folds
+  all,
+  any,
+
+  -- * Substrings
+
+  -- ** Breaking strings
+  take,
+  drop,
+  dropWhile,
+  dropWhileEnd,
+  breakEnd,
+  break,
+  span,
+  spanEnd,
+  splitAt,
+  split,
+  splitWith,
+  stripSuffix,
+
+  -- * Predicates
+  isInfixOf,
+  isPrefixOf,
+  isSuffixOf,
+
+  -- * Searching ShortByteStrings
+  --
+  -- ** Searching by equality
+  elem,
+
+  -- * Indexing ShortByteStrings
+  index,
+  elemIndex,
+  findIndex,
+) where
+
+import Prelude hiding
+    ( all
+    , any
+    , break
+    , concat
+    , drop
+    , dropWhile
+    , elem
+    , head
+    , init
+    , last
+    , length
+    , map
+    , null
+    , span
+    , splitAt
+    , tail
+    , take
+    )
+
+import AFP.Data.ByteString.Short
+    ( append, intercalate, isInfixOf, isPrefixOf, isSuffixOf, stripSuffix )
+import AFP.Data.ByteString.Short.Internal
+import AFP.Data.Word16
+
+import Data.ByteString.Short
+    ( ShortByteString
+    , empty
+    , fromShort
+    , index
+    , null
+    , packCString
+    , packCStringLen
+    , toShort
+    , useAsCString
+    , useAsCStringLen
+    )
+import Data.ByteString.Short.Internal
+    ( ShortByteString (SBS), createFromPtr )
+import qualified Data.Foldable as F
+import GHC.Exts
+import GHC.ST
+    ( ST (ST), runST )
+import GHC.Word
+
+import qualified Data.ByteString.Short as BS
+import qualified Data.ByteString.Short.Internal as BS
+import qualified Data.List as List
+
+
+-- -----------------------------------------------------------------------------
+-- Introducing and eliminating 'ShortByteString's
+
+-- | /O(1)/ Convert a 'Word16' into a 'ShortByteString'
+singleton :: Word16 -> ShortByteString
+singleton = \w -> create 2 (\mba -> writeWord16Array mba 0 w)
+{-# INLINE [1] singleton #-}
+
+
+-- | /O(n)/. Convert a list into a 'ShortByteString'
+pack :: [Word16] -> ShortByteString
+pack = packBytes
+
+
+-- | /O(n)/. Convert a 'ShortByteString' into a list.
+unpack :: ShortByteString -> [Word16]
+unpack = unpackBytes . assertEven
+
+
+-- ---------------------------------------------------------------------
+-- Basic interface
+
+length :: ShortByteString -> Int
+length = (`div` 2) . BS.length . assertEven
+
+infixr 5 `cons` --same as list (:)
+infixl 5 `snoc`
+
+-- | /O(n)/ Append a Word16 to the end of a 'ShortByteString'
+-- 
+-- Note: copies the entire byte array
+snoc :: Word16 -> ShortByteString -> ShortByteString
+snoc c = \(assertEven -> sbs) -> let l = length sbs
```

Comment:
We run this `assertEven` here everywhere. Is this a good idea? Does it affect performance? Do we pad zero-pad the input string if it isn't an even number of bytes instead?